tinynn/converter/operators/optimize.py

```diff
@@ -352,23 +352,37 @@ def fuse_conv2d_gather(self):
             # Update graph, prepare to drop the output tensor of the gather node and use the output tensor of the
             # conv instead
             conv['outputs'][0] = new_output
+            conv_out_quant_param = conv['op'].outputs[0].quantization
             conv['op'].outputs[0] = self.graph.tensor_map[new_output]
+            conv['op'].outputs[0].quantization = conv_out_quant_param
             self.graph.tensor_node_map[new_output] = conv['name']
             tensor['name'] = gather['outputs'][0]
             tensor['label'] = gather['outputs'][0]
             # permute weight of conv-op
             indx = gather['op'].inputs[1].tensor.copy()
             w = conv['op'].inputs[1].tensor.copy()
+            w_quant_param = conv['op'].inputs[1].quantization
             new_w = np.take(w, indx, axis=0)
-            new_w = self.create_attr_tensor(new_w)
+            if w_quant_param is not None and isinstance(w_quant_param.scale, list) and w_quant_param.dim == 0:
+                new_w_scale = np.take(w_quant_param.scale, indx, axis=0)
+                new_b_zeros = np.take(w_quant_param.zero_point, indx, axis=0)
+                w_quant_param.scale = new_w_scale
+                w_quant_param.zero_point = new_b_zeros
+            new_w = self.create_attr_tensor(new_w, quantization=w_quant_param)
             actions.append((self.graph.replace_operator_input, (conv, 1, new_w)))
 
             # permute bias of conv-op
             b = conv['op'].inputs[2].tensor.copy() if len(conv['op'].inputs) > 2 else None
+            b_quant_param = conv['op'].inputs[2].quantization
             new_b = None
             if b is not None:
                 new_b = np.take(b, indx, axis=0)
-                new_b = self.create_attr_tensor(new_b)
+                if b_quant_param is not None and isinstance(b_quant_param.scale, list) and b_quant_param.dim == 0:
```

Comment:
`b_quant_param.dim == 0` is not correct. Since it is always true (`b` has only one dimension), if `w_quant_param.dim !=0` then you are swapping `b` but not `w`, this is not right because they should be swapped together. 

Apparently, this correct logic is that when the condition in line 366 is satisfied, then we perform the logic in line 381-385.

Reply:
Thanks for the detailed comment.